### PR TITLE
Replace Version with String in YAML framework, pushing Version parsing where is actually needed

### PR DESCRIPTION
**Pull request by ldematte**
Remove un-needed call to `cat` API, replaced with the set of node versions we gather at test start anyway.
Keep node versions as a set of string-versions, push parsing "down" (closer to where it is used) and lazier - we parse only when a semantic version is really needed for comparison (`before` and `after`).

To become fully lazy, we need https://github.com/elastic/elasticsearch/pull/103404 too (see TODO)

### CODE REVIEW

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
@elasticmachine rerun elasticsearch-ci/8.10.0 / bwc

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
BwC tests are failing due to an unrelated issue (see https://github.com/elastic/elasticsearch/issues/103358 and others)

File: qa/multi-cluster-search/src/test/java/org/elasticsearch/search/MultiClusterSearchYamlTestSuiteIT.java [hunk 1]
```diff
@@ -13,52 +13,58 @@
...
 import java.util.function.Predicate;
 
 @TimeoutSuite(millis = 5 * TimeUnits.MINUTE) // to account for slow as hell VMs
 public class MultiClusterSearchYamlTestSuiteIT extends ESClientYamlSuiteTestCase {
 
-    private static Version remoteEsVersion = null;
+    private static String remoteEsVersion = null;
 
     @BeforeClass
     public static void determineRemoteClusterMinimumVersion() {
```
Comment (mosche) [inline qa/multi-cluster-search/src/test/java/org/elasticsearch/search/MultiClusterSearchYamlTestSuiteIT.java:33]: Unrelated to your change... wondering, this doesn't seem to do what it claims to do (finding the minimum version)? Is this correct?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java [hunk 2]
```diff
@@ -319,13 +312,15 @@ static Map<String, Set<Path>> loadSuites(String... paths) throws Exception {
...
                 Path path = root.resolve(strPath);
                 if (Files.isDirectory(path)) {
-                    Files.walk(path).forEach(file -> {
-                        if (file.toString().endsWith(".yml")) {
-                            addSuite(root, file, files);
-                        } else if (file.toString().endsWith(".yaml")) {
-                            throw new IllegalArgumentException("yaml files are no longer supported: " + file);
-                        }
-                    });
+                    try (var filesStream = Files.walk(path)) {
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java:315]: good catch 👍

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java [hunk 3]
```diff
@@ -376,8 +379,14 @@ public void execute(ClientYamlTestExecutionContext executionContext) throws IOEx
...
             // and feature check for v8. This way the version check can be removed once we move to v9
             @UpdateForV9
-            var fixedInV7 = executionContext.esVersion().major == Version.V_7_17_0.major
-                && executionContext.esVersion().onOrAfter(Version.V_7_17_2);
+            var fixedInV7 = executionContext.nodesVersions()
+                .stream()
+                .map(ESRestTestCase::parseLegacyVersion)
+                .flatMap(Optional::stream)
+                .min(VersionId::compareTo)
+                .map(v -> v.major == Version.V_7_17_0.major && v.onOrAfter(Version.V_7_17_2))
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java:387]: No strong preference, but prob easier to understand the intend using filter / isPresent instead 
```java
.filter(v -> v.major == Version.V_7_17_0.major && v.onOrAfter(Version.V_7_17_2))
.isPresent()
```

**Review by mosche — APPROVED** [review]
LGTM 🎉

**Comment by ldematte** [comment]
Thanks for the review @mosche, but I have decided to split part of this out to https://github.com/elastic/elasticsearch/pull/103626 in order to not do the same mistake I did this morning and break serverless submodule update. 

Plan is:
- https://github.com/elastic/elasticsearch/pull/103626
- Update serverless submodule, switch to the new overloads in serverless
- Update this PR (merge + remove the `@Deprecated` methods introduced in https://github.com/elastic/elasticsearch/pull/103626)

If you can review https://github.com/elastic/elasticsearch/pull/103626 too that would be great :)

**Comment by ldematte** [comment]
Removing full BwC as tests are still failing due to an unrelated issue (see https://github.com/elastic/elasticsearch/issues/103358 and others)

**Comment by ldematte** [comment]
@elasticsearchmachine please retest this

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
Compilation in serverless broken for another issue, should be fixed now.

@elasticmachine update branch

**Pull request merged**